bika/lims/catalog/catalog_multiplex_processor.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+
+from Acquisition import aq_base
+from bika.lims import api
+from bika.lims import logger
+from bika.lims.config import USE_COLLECTIVE_INDEXING
+from bika.lims.interfaces import IMultiCatalogBehavior
+from zope.interface import implements
+
+if USE_COLLECTIVE_INDEXING:
+    from collective.indexing.interfaces import IIndexQueueProcessor
+
+REQUIRED_CATALOGS = [
+    "auditlog_catalog",
+]
+
+
+class CatalogMultiplexProcessor(object):
+    """A catalog multiplex processor
+    """
+    if USE_COLLECTIVE_INDEXING:
+        implements(IIndexQueueProcessor)
+
+    def get_catalogs_for(self, obj):
+        catalogs = getattr(obj, "_catalogs", [])
+        for rc in REQUIRED_CATALOGS:
+            if rc in catalogs:
+                continue
+            catalogs.append(rc)
+        return map(api.get_tool, catalogs)
+
+    def supports_multi_catalogs(self, obj):
+        """Check if the Multi Catalog Behavior is enabled
+        """
+        if IMultiCatalogBehavior(obj, None) is None:
+            return False
+        return True
+
+    def index(self, obj, attributes=None):
+        if attributes is None:
+            attributes = []
+
+        if not self.supports_multi_catalogs(obj):
+            return
+
+        catalogs = self.get_catalogs_for(obj)
+        url = api.get_path(obj)
+
+        for catalog in catalogs:
+            logger.info(
+                "CatalogMultiplexProcessor::indexObject:catalog={} url={}"
+                .format(catalog.id, url))
+            # We want the intersection of the catalogs idxs
+            # and the incoming list.
+            indexes = set(catalog.indexes()).intersection(attributes)
```

Comment:
If attributes have been manually specified, but no matches with current catalog are found, the object will be reindexed for all indexes. I would rather skip this case for better performance:

```python
indexes = set(catalog.indexes()).intersection(attributes)
if attributes and not indexes:
    continue

logger.info(
    "CatalogMultiplexProcessor::indexObject:catalog={} url={}"
    .format(catalog.id, url))
catalog.catalog_object(obj, url, idxs=list(indexes))
```

Reply:
agree, will change that